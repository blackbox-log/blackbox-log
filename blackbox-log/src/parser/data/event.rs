use alloc::vec::Vec;

use tracing::instrument;

use crate::parser::{decode, ParseError, ParseResult, Reader};

#[derive(Debug, Clone, PartialEq)]
pub enum Event {
    SyncBeep(u64),
    InflightAdjustment {
        function: u8,
        new_value: AdjustedValue,
    },
    Resume {
        log_iteration: u32,
        time: u32,
    },
    Disarm(u32),
    FlightMode {
        flags: u32,
        last_flags: u32,
    },
    ImuFailure {
        error: u32,
    },
    End,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AdjustedValue {
    Float(f32),
    Int(i32),
}

impl Event {
    #[instrument(level = "debug", name = "Event::parse", skip_all, fields(kind))]
    pub(crate) fn parse_into(data: &mut Reader, events: &mut Vec<Self>) -> ParseResult<EventKind> {
        let byte = data.read_u8().ok_or(ParseError::UnexpectedEof)?;
        let kind = EventKind::from_byte(byte).unwrap_or_else(|| todo!("invalid event: {byte}"));

        match kind {
            EventKind::SyncBeep => {
                // TODO: SyncBeep handle time rollover

                let time = decode::variable(data)?;
                events.push(Self::SyncBeep(time.into()));
            }

            EventKind::InflightAdjustment => {
                let function = data.read_u8().ok_or(ParseError::UnexpectedEof)?;

                let new_value = if (function & 0x80) > 0 {
                    AdjustedValue::Float(data.read_f32().ok_or(ParseError::UnexpectedEof)?)
                } else {
                    AdjustedValue::Int(decode::variable_signed(data)?)
                };

                events.push(Self::InflightAdjustment {
                    function: function & 0x7F,
                    new_value,
                });
            }

            EventKind::Resume => {
                let log_iteration = decode::variable(data)?;
                let time = decode::variable(data)?;

                events.push(Self::Resume {
                    log_iteration,
                    time,
                });
            }

            EventKind::Disarm => {
                let reason = decode::variable(data)?;
                events.push(Self::Disarm(reason));
            }

            EventKind::FlightMode => {
                let flags = decode::variable(data)?;
                let last_flags = decode::variable(data)?;
                events.push(Self::FlightMode { flags, last_flags });
            }

            EventKind::ImuFailure => {
                let error = decode::variable(data)?;
                events.push(Self::ImuFailure { error });
            }

            EventKind::End => {
                check_message(data, b"End of log")?;

                if data.peek() == Some(b' ') {
                    // Assume INAV's new format:
                    // `End of log (disarm reason:x)\0`

                    check_message(data, b" (disarm reason:")?;

                    let reason = data.read_u8().ok_or(ParseError::UnexpectedEof)?;
                    events.push(Self::Disarm(reason.into()));

                    if data.read_u8() != Some(b')') {
                        return Err(ParseError::Corrupted);
                    }
                }

                if data.read_u8() != Some(0) {
                    return Err(ParseError::Corrupted);
                }

                events.push(Self::End);
            }
        }

        Ok(kind)
    }
}

byte_enum! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    #[repr(u8)]
    pub(crate) enum EventKind {
        SyncBeep = 0,
        InflightAdjustment = 13,
        Resume = 14,
        Disarm = 15,
        FlightMode = 30,
        ImuFailure = 40,
        End = 255,
    }
}

fn check_message(bytes: &mut Reader, message: &[u8]) -> ParseResult<()> {
    let bytes = bytes.read_n_bytes(message.len());

    if bytes.len() != message.len() {
        return Err(ParseError::UnexpectedEof);
    }

    if bytes != message {
        return Err(ParseError::Corrupted);
    }

    Ok(())
}
