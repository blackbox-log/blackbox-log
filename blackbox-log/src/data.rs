//! Types for the data section of blackbox logs.

use alloc::vec::Vec;

use crate::event::{self, Event};
use crate::frame::{DataFrameKind, FrameKind, GpsFrame, GpsHomeFrame, MainFrame, SlowFrame};
use crate::parser::InternalError;
use crate::{Headers, Reader};

#[derive(Debug, Clone)]
pub(crate) struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<FrameSync>,
    pub(crate) slow_frames: Vec<SlowFrame>,
    pub(crate) gps_frames: Vec<GpsFrame>,
    pub(crate) gps_home_frames: Vec<GpsHomeFrame>,
}

#[derive(Debug, Clone)]
pub(crate) struct FrameSync {
    pub(crate) main: MainFrame,
    pub(crate) slow: usize,
    pub(crate) gps: Option<usize>,
}

impl FrameSync {
    pub(crate) const fn new(main: MainFrame, slow: &[SlowFrame], gps: &[GpsFrame]) -> Self {
        Self {
            main,
            slow: slow.len() - 1,
            gps: gps.len().checked_sub(1),
        }
    }
}

/// Statistics about a decoded log.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[non_exhaustive]
pub struct Stats {
    /// The number of valid frames found of each type.
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
    pub gps: usize,
    pub gps_home: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
                gps: self.gps_frames.len(),
                gps_home: self.gps_home_frames.len(),
            },
        }
    }

    pub(crate) fn parse(data: &mut Reader, headers: &Headers) -> Self {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        let mut slow_frames = Vec::new();
        let mut gps_frames = Vec::new();
        let mut gps_home_frames = Vec::new();

        slow_frames.push(headers.slow_frames.empty_frame());

        if let Some(def) = &headers.gps_frames {
            gps_frames.push(def.empty_frame());
        }

        let mut restore;
        let mut last_kind = None;
        while let Some(byte) = data.read_u8() {
            restore = data.get_restore_point();

            let Some(kind) = FrameKind::from_byte(byte) else {
                tracing::debug!("found invalid frame byte: {byte:0>#2x}");

                if let Some(last_kind) = last_kind.take() {
                    data.restore(restore);

                    match last_kind {
                        FrameKind::Event => {
                            events.pop();
                        }
                        FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                            main_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Slow) => {
                            slow_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Gps) => {
                            gps_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::GpsHome) => {
                            gps_home_frames.pop();
                        }
                    };
                }

                skip_to_frame(data);
                continue;
            };

            tracing::trace!("trying to parse {kind:?} frame");

            let result = match kind {
                FrameKind::Event => match Event::parse_into(data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                    let frame = MainFrame::parse(data, kind, &main_frames, headers);
                    frame.map(|frame| {
                        main_frames.push(FrameSync::new(frame, &slow_frames, &gps_frames));
                    })
                }
                FrameKind::Data(DataFrameKind::Gps) => headers.gps_frames.as_ref().map_or_else(
                    || {
                        tracing::debug!("found GPS frame without GPS frame definition");
                        Err(InternalError::Retry)
                    },
                    |gps| {
                        gps.parse(
                            data,
                            headers,
                            main_frames.last().map(|sync| &sync.main),
                            gps_home_frames.last(),
                        )
                        .map(|frame| gps_frames.push(frame))
                    },
                ),
                FrameKind::Data(DataFrameKind::GpsHome) => {
                    headers.gps_home_frames.as_ref().map_or_else(
                        || {
                            tracing::debug!(
                                "found GPS home frame without GPS home frame definition"
                            );
                            Err(InternalError::Retry)
                        },
                        |gps_home| {
                            gps_home
                                .parse(data, headers)
                                .map(|frame| gps_home_frames.push(frame))
                        },
                    )
                }
                FrameKind::Data(DataFrameKind::Slow) => headers
                    .slow_frames
                    .parse(data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(InternalError::Retry) => {
                    tracing::debug!("found corrupted {kind:?} frame");
                    data.restore(restore);
                    skip_to_frame(data);
                }
                Err(InternalError::Eof) => {
                    tracing::debug!("found unexpected end of file in data section");
                    break;
                }
            }
        }

        Self {
            events,
            main_frames,
            slow_frames,
            gps_frames,
            gps_home_frames,
        }
    }
}

#[cold]
fn skip_to_frame(data: &mut Reader) {
    data.skip_until_any(
        &[
            FrameKind::Event,
            FrameKind::Data(DataFrameKind::Intra),
            FrameKind::Data(DataFrameKind::Slow),
            FrameKind::Data(DataFrameKind::Gps),
            FrameKind::Data(DataFrameKind::GpsHome),
        ]
        .map(u8::from),
    );
}
