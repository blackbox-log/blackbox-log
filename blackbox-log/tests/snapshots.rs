use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::{ParseEvent, Stats};
use blackbox_log::event::Event;
use blackbox_log::units::{si, Flag, FlagSet};
use blackbox_log::{DataParser, Headers, Unit, Value};
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let logs = file
                .iter()
                .map(|mut reader| {
                    Headers::parse(&mut reader).map(|headers| {
                        let data = DataParser::new(&mut reader, &headers);
                        LogSnapshot::new(&headers, data)
                    })
                })
                .collect::<Vec<_>>();

            insta::assert_ron_snapshot!(logs);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'data> {
    headers: Headers<'data>,
    stats: Stats,
    events: Vec<Event>,
    main: Fields,
    slow: Fields,
    gps: Fields,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_, '_>) -> Self {
        let headers = headers.clone();

        let mut events = Vec::new();
        let mut main = headers.main_def().iter().collect::<Fields>();
        let mut slow = headers.slow_def().iter().collect::<Fields>();
        let mut gps = headers
            .gps_def()
            .iter()
            .flat_map(|def| def.iter())
            .collect::<Fields>();

        while let Some(frame) = data.next() {
            match frame {
                ParseEvent::Event(event) => events.push(event),
                ParseEvent::Main(frame) => main.update(frame),
                ParseEvent::Slow(frame) => slow.update(frame),
                ParseEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers,
            stats: data.stats().clone(),
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<T, U> FromIterator<(T, U)> for Fields
where
    T: Into<String>,
    U: Into<Unit>,
{
    fn from_iter<I: IntoIterator<Item = (T, U)>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|(name, unit)| FieldSnapshot::new(name.into(), unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::FrameTime
                | Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::FrameTime(t) => t.get::<si::time::microsecond>().round() as i128,
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
