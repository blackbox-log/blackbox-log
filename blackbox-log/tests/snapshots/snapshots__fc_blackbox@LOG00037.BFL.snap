---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: blackbox-log/tests/logs/.fc-blackbox/src/test-data/LOG00037.BFL
---
[
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "Betaflight 4.2.0 (8f2d21460) STM32F745",
      firmware_kind: Betaflight,
      board_info: Some("HBRO KAKUTEF7"),
      craft_name: Some("AR8"),
      vbat: Some(VbatConfig(
        reference: 2277,
        scale: 110,
      )),
      current_meter: Some(CurrentMeterConfig(
        offset: 0,
        scale: 275,
      )),
      acceleration_1g: Some(2048),
      gyro_scale: Some(0.017453292),
      min_throttle: Some(1070),
      motor_output_range: Some(MotorOutputRange(158, 2047)),
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 16774,
        slow: 4,
        gps_home: 1,
      ),
    ),
    events: [
      SyncBeep(451840837),
      Disarm(4),
      End,
    ],
    fields: Fields([
      FieldSnapshot(
        name: "loopIteration",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 134149,
          unique: 16774,
          histogram: (1049, 1049, 1049, 1049, 1049, 1049, 1048, 1048, 1048, 1048, 1048, 1048, 1048, 1048, 1048, 1048),
        ),
      ),
      FieldSnapshot(
        name: "time",
        unit: FrameTime,
        history: FieldSnapshot(
          min: 0,
          max: 469230773,
          unique: 16774,
          histogram: (1070, 1082, 1113, 945, 1091, 677, 1378, 1050, 1048, 1050, 1119, 951, 1102, 684, 1362, 1052),
        ),
      ),
      FieldSnapshot(
        name: "axisP[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -49,
          max: 80,
          unique: 89,
          histogram: (3977, 3227, 2530, 1824, 1341, 960, 672, 541, 367, 292, 243, 241, 192, 184, 117, 66),
        ),
      ),
      FieldSnapshot(
        name: "axisP[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -583,
          max: 284,
          unique: 198,
          histogram: (3662, 2960, 2170, 1506, 1171, 918, 826, 697, 614, 532, 421, 398, 316, 223, 181, 179),
        ),
      ),
      FieldSnapshot(
        name: "axisP[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -73,
          max: 23,
          unique: 72,
          histogram: (6030, 4391, 2658, 1541, 916, 496, 243, 160, 95, 69, 49, 42, 32, 14, 22, 16),
        ),
      ),
      FieldSnapshot(
        name: "axisI[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -50,
          max: 11,
          unique: 62,
          histogram: (1888, 864, 1203, 1858, 1494, 1409, 1772, 1633, 1183, 799, 506, 377, 347, 416, 483, 542),
        ),
      ),
      FieldSnapshot(
        name: "axisI[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -251,
          max: 92,
          unique: 195,
          histogram: (1567, 839, 1346, 1352, 1610, 2571, 1943, 1133, 1053, 942, 765, 460, 403, 306, 258, 226),
        ),
      ),
      FieldSnapshot(
        name: "axisI[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -61,
          max: 44,
          unique: 101,
          histogram: (2713, 1770, 2205, 1936, 1969, 1962, 1453, 906, 614, 476, 231, 146, 131, 104, 64, 94),
        ),
      ),
      FieldSnapshot(
        name: "axisD[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -122,
          max: 192,
          unique: 106,
          histogram: (3422, 3001, 2709, 2042, 1647, 1219, 864, 605, 405, 274, 181, 124, 95, 72, 60, 54),
        ),
      ),
      FieldSnapshot(
        name: "axisD[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -893,
          max: 526,
          unique: 140,
          histogram: (3571, 3196, 2698, 2045, 1537, 1019, 742, 592, 384, 265, 225, 163, 137, 85, 69, 46),
        ),
      ),
      FieldSnapshot(
        name: "axisF[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -1,
          max: 3,
          unique: 5,
          histogram: (16419, 312, 23, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisF[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -2,
          max: 2,
          unique: 5,
          histogram: (16269, 436, 69, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisF[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -7,
          max: 1,
          unique: 9,
          histogram: (16539, 190, 9, 5, 4, 7, 6, 14, 0, 0, 0, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -85,
          max: 98,
          unique: 184,
          histogram: (7414, 1242, 544, 800, 613, 687, 543, 740, 459, 576, 617, 375, 495, 652, 456, 561),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -76,
          max: 111,
          unique: 188,
          histogram: (985, 1079, 802, 6789, 1319, 736, 625, 703, 404, 426, 512, 361, 500, 675, 405, 453),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -96,
          max: 78,
          unique: 175,
          histogram: (1443, 4241, 1232, 1963, 1109, 883, 532, 1212, 382, 529, 531, 402, 627, 880, 407, 401),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1948,
          unique: 754,
          histogram: (950, 850, 794, 582, 880, 1324, 1449, 765, 2428, 798, 642, 1286, 1288, 986, 1004, 748),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -27,
          max: 31,
          unique: 59,
          histogram: (7432, 1193, 766, 794, 914, 1184, 696, 694, 702, 494, 374, 452, 345, 197, 292, 245),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -23,
          max: 36,
          unique: 60,
          histogram: (1359, 8165, 935, 851, 907, 857, 890, 535, 486, 322, 264, 225, 225, 218, 313, 222),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -35,
          max: 28,
          unique: 64,
          histogram: (4666, 3142, 1700, 827, 1018, 856, 562, 972, 673, 351, 248, 497, 393, 253, 283, 333),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 965,
          unique: 729,
          histogram: (2417, 759, 774, 1271, 1280, 956, 958, 686, 926, 823, 796, 690, 911, 1293, 1345, 889),
        ),
      ),
      FieldSnapshot(
        name: "vbatLatest",
        unit: Voltage,
        history: FieldSnapshot(
          min: 0,
          max: 20220,
          unique: 140,
          histogram: (1832, 393, 1753, 965, 787, 1074, 771, 1347, 856, 1300, 615, 1262, 803, 1242, 1360, 414),
        ),
      ),
      FieldSnapshot(
        name: "amperageLatest",
        unit: Amperage,
        history: FieldSnapshot(
          min: 0,
          max: 351004,
          unique: 591,
          histogram: (1621, 1001, 1342, 828, 1145, 1004, 1052, 922, 986, 947, 927, 885, 1126, 1109, 949, 930),
        ),
      ),
      FieldSnapshot(
        name: "magADC[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -740,
          max: 292,
          unique: 117,
          histogram: (1282, 791, 1282, 1677, 984, 786, 1185, 493, 1084, 591, 592, 1480, 1427, 984, 852, 1284),
        ),
      ),
      FieldSnapshot(
        name: "magADC[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -66,
          max: 947,
          unique: 126,
          histogram: (1480, 395, 985, 888, 1382, 1775, 393, 982, 1679, 951, 1476, 987, 689, 1184, 639, 889),
        ),
      ),
      FieldSnapshot(
        name: "magADC[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 2576,
          unique: 123,
          histogram: (1187, 1281, 1084, 788, 493, 690, 1084, 1133, 885, 1085, 1149, 1579, 1084, 1182, 788, 1282),
        ),
      ),
      FieldSnapshot(
        name: "BaroAlt",
        unit: Unitless,
        history: FieldSnapshot(
          min: -656,
          max: 289,
          unique: 255,
          histogram: (914, 419, 1376, 979, 1014, 716, 1506, 2241, 690, 1057, 689, 839, 1078, 568, 1652, 1036),
        ),
      ),
      FieldSnapshot(
        name: "rssi",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1023,
          unique: 1,
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[0]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -30,
          max: 37,
          unique: 68,
          histogram: (3820, 3984, 1450, 956, 904, 827, 664, 694, 672, 574, 486, 512, 415, 386, 236, 194),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[1]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -120,
          max: 197,
          unique: 136,
          histogram: (3222, 4965, 1926, 1212, 953, 853, 674, 705, 501, 361, 324, 194, 254, 317, 134, 179),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[2]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -43,
          max: 28,
          unique: 72,
          histogram: (3436, 3829, 1941, 1067, 995, 946, 739, 901, 532, 342, 394, 391, 383, 347, 280, 251),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[0]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -716,
          max: 566,
          unique: 387,
          histogram: (1225, 1347, 1365, 1242, 1167, 1179, 1050, 1114, 911, 921, 848, 1041, 817, 874, 876, 797),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[1]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -414,
          max: 34,
          unique: 222,
          histogram: (923, 914, 881, 921, 1025, 937, 1047, 990, 1171, 1135, 1131, 1371, 1123, 1119, 1023, 1063),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[2]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: 0,
          max: 3614,
          unique: 1680,
          histogram: (1014, 1025, 993, 1109, 1029, 1032, 1136, 1042, 1267, 1083, 1062, 976, 995, 1004, 1007, 1000),
        ),
      ),
      FieldSnapshot(
        name: "debug[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -33,
          max: 45,
          unique: 75,
          histogram: (1936, 3360, 2413, 1737, 1282, 1004, 844, 693, 642, 568, 541, 477, 373, 356, 294, 254),
        ),
      ),
      FieldSnapshot(
        name: "debug[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -125,
          max: 214,
          unique: 141,
          histogram: (2061, 3251, 2580, 1998, 1631, 1159, 906, 702, 534, 392, 362, 301, 294, 224, 222, 157),
        ),
      ),
      FieldSnapshot(
        name: "debug[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -54,
          max: 33,
          unique: 83,
          histogram: (1684, 2803, 2474, 1983, 1642, 1230, 902, 707, 604, 544, 466, 435, 380, 329, 307, 284),
        ),
      ),
      FieldSnapshot(
        name: "debug[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 0,
          unique: 1,
        ),
      ),
      FieldSnapshot(
        name: "motor[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 2047,
          unique: 1172,
          histogram: (1017, 999, 1034, 981, 1050, 1026, 1033, 1022, 1030, 1069, 1075, 1066, 1133, 1045, 1112, 1082),
        ),
      ),
      FieldSnapshot(
        name: "motor[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 2044,
          unique: 1193,
          histogram: (1078, 1056, 985, 1017, 983, 1077, 1030, 1067, 1009, 989, 1020, 1103, 1023, 1192, 1083, 1062),
        ),
      ),
      FieldSnapshot(
        name: "motor[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1901,
          unique: 1153,
          histogram: (1057, 1042, 1023, 1079, 1035, 1038, 979, 1029, 1048, 1058, 1051, 1037, 1096, 1126, 1069, 1007),
        ),
      ),
      FieldSnapshot(
        name: "motor[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 2052,
          unique: 1190,
          histogram: (1032, 1067, 1018, 1055, 1030, 1063, 997, 1030, 1048, 1040, 1044, 1088, 1201, 1049, 1013, 999),
        ),
      ),
      FieldSnapshot(
        name: "flightModeFlags",
        unit: FlightMode,
        history: {
          "Arm": 16773,
          "Blackbox": 16773,
        },
      ),
      FieldSnapshot(
        name: "stateFlags",
        unit: State,
        history: {
          "GpsFix": 16773,
          "GpsFixHome": 16773,
        },
      ),
      FieldSnapshot(
        name: "failsafePhase",
        unit: FailsafePhase,
        history: {},
      ),
      FieldSnapshot(
        name: "rxSignalReceived",
        unit: Boolean,
        history: History(
          yes: 16773,
          no: 1,
        ),
      ),
      FieldSnapshot(
        name: "rxFlightChannelsValid",
        unit: Boolean,
        history: History(
          yes: 16773,
          no: 1,
        ),
      ),
    ]),
  )),
]
